File: problems/0601-0700/0621_shortest-cooldown-schedule/solution_heap_timetable.java
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.HashMap;
import java.util.Map;
import java.util.PriorityQueue;

class Solution {

    public int shortestCooldownSchedule(String[] jobs, int n) {
        Map<String, Integer> counts = new HashMap<>();
        for (String t : jobs) {
            counts.merge(t, 1, Integer::sum);
        }
        // Max-heap of remaining counts for labels free to run right now; only
        // the counts matter, because the cooldown rule treats every label
        // alike.
        PriorityQueue<Integer> ready = new PriorityQueue<>((a, b) -> Integer.compare(b, a));
        for (int v : counts.values()) {
            ready.add(v);
        }
        // FIFO of runs still cooling: [slot when the label may run again,
        // remaining count]. Free slots arrive in order, so the front pops.
        Deque<int[]> cooling = new ArrayDeque<>();
        int time = 0;
        while (!ready.isEmpty() || !cooling.isEmpty()) {
            // Release everything whose cooldown has expired by now.
            while (!cooling.isEmpty() && cooling.peek()[0] <= time) {
                ready.add(cooling.poll()[1]);
            }
            if (ready.isEmpty()) {
                // Nothing can run: jump the clock straight to the next
                // release instead of counting idle slots one by one.
                time = cooling.peek()[0];
                continue;
            }
            // Run one job of the largest remaining count.
            int top = ready.poll();
            if (top > 1) {
                cooling.add(new int[] { time + n + 1, top - 1 });
            }
            time++;
        }
        return time;
    }
}

File: problems-originals/0601-0700/0621_task-scheduler/solution_heap_timetable.py
import heapq
from collections import Counter, deque


class Solution:
    def leastInterval(self, tasks: list[str], n: int) -> int:
        counts = Counter(tasks)
        # Max-heap (negated counts, since heapq is a min-heap) of every label
        # free to run right now; only the counts matter, because the cooldown
        # rule treats every label alike.
        ready = [-c for c in counts.values()]
        heapq.heapify(ready)
        # FIFO of runs still cooling: (slot when the label may run again,
        # negated count left). Free slots arrive in order, so the front pops.
        cooling = deque()
        time = 0
        while ready or cooling:
            # Release everything whose cooldown has expired by now.
            while cooling and cooling[0][0] <= time:
                heapq.heappush(ready, cooling.popleft()[1])
            if not ready:
                # Nothing can run: jump the clock straight to the next
                # release instead of counting idle slots one by one.
                time = cooling[0][0]
                continue
            # Run one job of the largest remaining count.
            top = -heapq.heappop(ready)
            if top > 1:
                cooling.append((time + n + 1, -(top - 1)))
            time += 1
        return time
